lib/tools/apk-utils.js

```diff
@@ -456,34 +442,87 @@ apkUtilsMethods.installOrUpgrade = async function (appPath, pkg = null, options
-      return;
+      log.debug(`The version code of the installed '${pkg}' is greater than the application version code (${pkgVersionCode} > ${apkVersionCode})`);
+      return this.APP_INSTALL_STATE.NEWER_VERSION_INSTALLED;
     }
     // Version codes might not be maintained. Check version names.
     if (pkgVersionCode === apkVersionCode) {
       if (_.isString(apkVersionName) && _.isString(pkgVersionName) && semver.satisfies(pkgVersionName, `>=${apkVersionName}`)) {
-        log.debug(`The installed '${pkg}' package does not require upgrade ('${pkgVersionName}' >= '${apkVersionName}')`);
-        return;
+        log.debug(`The version name of the installed '${pkg}' is greater or equal to the application version name ('${pkgVersionName}' >= '${apkVersionName}')`);
+        return this.APP_INSTALL_STATE.NEWER_VERSION_INSTALLED;
       }
       if (!_.isString(apkVersionName) || !_.isString(pkgVersionName)) {
-        log.debug(`The installed '${pkg}' package does not require upgrade (${pkgVersionCode} === ${apkVersionCode})`);
-        return;
+        log.debug(`The version name of the installed '${pkg}' is equal to application version name (${pkgVersionCode} === ${apkVersionCode})`);
+        return this.APP_INSTALL_STATE.SAME_VERSION_INSTALLED;
       }
     }
   } else if (_.isString(apkVersionName) && _.isString(pkgVersionName) && semver.satisfies(pkgVersionName, `>=${apkVersionName}`)) {
-    log.debug(`The installed '${pkg}' package does not require upgrade ('${pkgVersionName}' >= '${apkVersionName}')`);
-    return;
+    log.debug(`The version name of the installed '${pkg}' is greater or equal to the application version name ('${pkgVersionName}' >= '${apkVersionName}')`);
+    return this.APP_INSTALL_STATE.NEWER_VERSION_INSTALLED;
+  }
+
+  log.debug(`The installed '${pkg}' package is older than '${appPath}' (${pkgVersionCode} < ${apkVersionCode} or '${pkgVersionName}' < '${apkVersionName}')'`);
+  return this.APP_INSTALL_STATE.OLDER_VERSION_INSTALLED;
+};
+
+/**
+ * @typedef {Object} InstallOrUpgradeOptions
+ * @property {number} timeout [60000] - The count of milliseconds to wait until the
+ *                                      app is installed.
+ * @property {boolean} allowTestPackages [false] - Set to true in order to allow test
+ *                                                 packages installation.
+ * @property {boolean} useSdcard [false] - Set to true to install the app on sdcard
+ *                                         instead of the device memory.
+ * @property {boolean} grantPermissions [false] - Set to true in order to grant all the
+ *                                                permissions requested in the application's manifest
+ *                                                automatically after the installation is completed
+ *                                                under Android 6+.
+ */
+
+/**
+ * Install the package from the local file system of upgrade it if an older
```

Comment:
s/the local file system of upgrade it/the local file system or upgrade it/ ? (_of_ => _or_)